### ISSUE TITLE
(drive) If you put a shortcut to FolderX inside FolderX rclone will recursivly download FoulderX inside itself until you run out of disk space

### ISSUE THREAD

**Issue by rt4mn** [rclone/rclone#7118]
#### The associated forum post URL from `https://forum.rclone.org`

NA

#### What is the problem you are having with rclone?

When using rclone with google drive as the source, if you dont use `--drive-skip-shortcuts` rclone wont create a shortcut on the other end, it will just reacreat the files. this is an issue when you have a shortcut to foulderx *inside foulderx*. You get a balooning recursive file copies. So if on gdrive you have:

```
- Foulder1
  - file.txt
  - file2.txt
  - shortcut to Foulder1
```

and you run `rclone sync drive:Foulder1 local:Foulder1`

on local machine you will see:

```
- Foulder1
  - file.txt
  - file2.txt
  - Foulder1
    - file.txt
    - file2.txt
    - Foulder1
      - file.txt
      - file2.txt
      - Foulder1
        - file.txt
```

and so on and so on until you run out of disk space (if copying to local) or upload limit (if syncing to another gdrive). 

#### What is your rclone version (output from `rclone version`)

rclone v1.62.2

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

debian 11, 64

#### Which cloud storage system are you using? (e.g. Google Drive)

drive and linux

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`rclone sync drive:Foulder1 local:Foulder1`

and

`rclone sync drive:Foulder1 drive2:Foulder1`

both result in recursive foulders if 

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

NA. 

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
My first thought was to say "don't do that," but then I saw we have users in the equation :laughing: 

Rclone could keep a list of shortcut IDs it has traversed and the path it found them at. If it ever found a shortcut ID with a different path it could then assume there has been a recursive directory loop.

Rclone could just leave the shortcut out of the directory listing at this point and things will carry on just fine after that. It should probably write an ERROR log warning you about things.

How does this sound?

In an ideal world rclone would re-create the shortcuts as symlinks. However they don't have the right semantics to be symlinks as they don't link to a given path they link to a folder or file ID which could be anywhere (rclone doesn't know and there is no way of finding out).

**Comment by rt4mn** [comment]
> My first thought was to say "don't do that," but then I saw we have users in the equation laughing

Yeah, when I saw that my 4tb drive had filled up when backing up a 2tb was very confused, and when I saw the cause I was like "why would anybody do that?" but I think its understandable for heavy drive users making a  bunch of shortcuts to misfile a couple, and I know I've got a bunch of symlinks that lead noware in my local filesystem.

> How does this sound?

Seems fine to me? Honestly this bug(?) is not affecting my use case anymore since I got the go-ahead to [skip shortcuts entirely](https://forum.rclone.org/t/is-it-possible-to-do-a-google-drive-copy-that-recreates-shortcuts/31215/2), so I'm not really in a position to say, just wanted to raise this as a potential issue for future normal folks trying to migrate drive data.

**Comment by lamongabriel** [comment]
+1 to this - I just had a case where a user added a shortcut inside another shortcut, and our pipeline ran out of disk space.

For now, we're just skipping shortcuts entirely

**Comment by vscfreire** [comment]
Hello all,
 
I am also having this problem, the only difference being that I am performing a mount instead of a sync. Is this infinite recursion a problem that is going to be addressed in the future or is the solution to completely skip all shortcuts? It seems like the latter is undesirable, at least in my use case, since I want to access "Shared With Me" folders which are configured as shortcuts in "My Drive".
 
Thanks.

**Comment by ncw** [comment]
I had an idea about this - we already have the ancestor IDs in the directory cache, so we can fix this without using any more memory and at a very small number of lines of code.

Please give this a go

[v1.75.0-beta.9851.00bf6c19d.fix-7118-drive-shortcut-loop](https://beta.rclone.org/branch/fix-7118-drive-shortcut-loop/v1.75.0-beta.9851.00bf6c19d.fix-7118-drive-shortcut-loop/) on branch [fix-7118-drive-shortcut-loop](https://github.com/rclone/rclone/tree/fix-7118-drive-shortcut-loop) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've merged a fix for this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.75
